### DISCUSSION TITLE
GET API executing during build time but not in the run time

### DISCUSSION THREAD

**Discussion by jyoti-r-4** [vercel/next.js#68357] [App Router]
I'm woking on Next.js 14.2.3. I'm writting my API inside `app/api/v1/variable/abc/route.ts` . I've written a GET() like below:
```
import { NextResponse } from "next/server";
import { CONSTANT } from "../../../../utility/constant";
import { IAPIRequest } from "../../../../utility/type";
import { api, responseHandler } from "../../../../utility/utils";
import { AxiosError } from "axios";

export async function GET(){
    try{
        const url = process.env.TEMPLATE_MANAGER_SERVICE_BASE_URL +
        CONSTANT.templateManagerUrl +
        CONSTANT.variableUrl +
        CONSTANT.variableTypeUrl

        const request: IAPIRequest = {
            method: 'GET',
            url:url
        }

        const response = await api(request);
        console.log(response.data);

        if(!response.data || response.data.length === 0){
            return NextResponse.json({message:'No Content'}, {status:204})
        }

        return NextResponse.json(response.data, {status: 200})
    }catch(error){
        if(error instanceof AxiosError){
            if(error.response){
                const errorResponse = responseHandler(error.response);
                return NextResponse.json(errorResponse)
            }else{
                return NextResponse.json({
                    statusCode:500,
                    error: CONSTANT.internalError
                })
            }
        }else{
            return NextResponse.json(
                {error: 'An unexpected error occurred'},
                {status: 500}
            )
        }
    }
}
```

During build time, it's printing the log which is `response.data` in my code. But when the code is running during production i.e. `npm start`, it's not giving me updated data from the database. But it's giving the data which was present during the build time.

But if I'll change the method to POST, it's working fine as expected.

I've written other GET() in the same way but those are not considered as static, but this one only considered as static. Why?

Is there any issue with the GET method? Can someone help here?

This screenshot is for your reference:
![image](https://github.com/user-attachments/assets/d4d990ab-c606-49ce-a1b1-8ef60fe7472b)

**Comment by icyJoseph** [comment]
There's something called, automatic cache optimization. This is rather controversial, so much so that, Next 15, will drop it.

- https://nextjs.org/docs/app/building-your-application/routing/route-handlers#opting-out-of-caching

You can just add a `export const dynamic = 'force-dynamic'` to your file, and that'll opt-out from caching.

- A bit more about caching, https://nextjs.org/docs/app/building-your-application/routing/route-handlers#caching
- Next 15 will drop this automatic caching, https://nextjs.org/blog/next-15-rc#get-route-handlers-are-no-longer-cached-by-default

**Comment by jyoti-r-4** [comment]
Thanks Joseph. As you said, I've added `export const dynamic = 'force-dynamic'` at the end of the route.ts file and built. It's giving the same result. Correct me if I'm wrong.

**Reply by icyJoseph** [reply]
Is it still showing it as statically generated? And not with an `f`?

**Reply by jyoti-r-4** [reply]
> Thanks Joseph. As you said, I've added `export const dynamic = 'force-dynamic'` at the end of the route.ts file and built. It's giving the same result. Correct me if I'm wrong.

Yes, still it persists.

**Reply by icyJoseph** [reply]
Show me the code and build output

**Reply by jyoti-r-4** [reply]
```
import { NextResponse } from "next/server";
import { CONSTANT } from "../../../../utility/constant";
import { IAPIRequest } from "../../../../utility/type";
import { api, responseHandler } from "../../../../utility/utils";
import { AxiosError } from "axios";

export async function GET(){
    try{
        const url = process.env.TEMPLATE_MANAGER_SERVICE_BASE_URL +
        CONSTANT.templateManagerUrl +
        CONSTANT.variableUrl +
        CONSTANT.variableTypeUrl

        const request: IAPIRequest = {
            method: 'GET',
            url:url
        }

        const response = await api(request);
        console.log(response.data);

        if(!response.data || response.data.length === 0){
            return NextResponse.json({message:'No Content'}, {status:204})
        }

        return NextResponse.json(response.data, {status: 200})
    }catch(error){
        if(error instanceof AxiosError){
            if(error.response){
                const errorResponse = responseHandler(error.response);
                return NextResponse.json(errorResponse)
            }else{
                return NextResponse.json({
                    statusCode:500,
                    error: CONSTANT.internalError
                })
            }
        }else{
            return NextResponse.json(
                {error: 'An unexpected error occurred'},
                {status: 500}
            )
        }
    }
}
export const dynamic = 'force-dynamic'
```

This is my updated code.

**Reply by icyJoseph** [reply]
What's the path to that file?

**Reply by jyoti-r-4** [reply]
> What's the path to that file?

Path is: `app/api/v1/variable/variable-types/route.ts`

**Reply by icyJoseph** [reply]
🤔 really odd that it is doing this... maybe try to do something like:

```js
export const dynamic = 'force-dynamic';

export async function GET(request) {
  const url = new URL(request.url)
  console.log(url)
  // rest of your stuff
```

**Reply by jyoti-r-4** [reply]
No @icyJoseph, this is working when I'm adding
```
const url = new URL(request.url)
console.log(url)
```
Otherwise it's not working.
But to work around this, I changed the method to POST and it's executing as expected.
